added child device proposal to config management (#1236)
thin-edge/thin-edge.io: docs/src/references/child-devices.md

   3) associate consumed/provided data of a child-device (external or logical) to it's child-device twin in the cloud

### (1) Child-Device Provisioning

A newly attached child-device (external or logical) must be once provisioned. The provisioning phase of a child-device includes:
  1) creating cloud's child-device twin
  2) declaring all supported capabilities of the child-device to its child device twin
     (i.E. declaring all _supported operations_ and all _types_ per supported operation, as _config types_, _log types_, ...)
     (i.e. declaring all operations supported by the device like configuration management, software management etc and further metadata per supported operation, like _config types_, _log types_, _software list_ etc)

There are two option to create the cloud's child-device twin and declaring all supported capabilities. 

On the one hand a cloud's _child-device twin_ can be created upfront (e.g. with some customer-specific cloud-site backend), and also all capabilities can be declared that way to the new twin. Then the child-device on the device site (external or logical) relies on that existing twin.

> On the other hand a child-device (external or logical) can announce it's capabilities to thin-edge to provoke thin-edge creating the cloud's child-device twin and declaring all capabilities on that child-device twin. 

Reviewer: ```suggestion
On the other hand, a child-device (external or logical) can announce its capabilities to the thin-edge device to trigger the thin-edge for creating the child-device twin in the cloud and declare all the capabilities on that child-device twin.
```

##### Announcing child-device capabilities by the child-device it-self

To announce it's capabilities the child device (external or logical) sends an MQTT message to `tedge/meta/<childid>`. The MQTT message is as below:

```json
{
   "device-name": "<optional name of that child-device>",
   "device-type": "<optional type of that child-device>",
   "<capability 1>": <capability specific JSON object>,
   "<capability 2>": <capability specific JSON object>,
   // [...]
}
```

Thereby the fields are as below:

>    * `device-name` is an optional human readble device-name, visible in the cloud. If the field is not contained in the JSON message the `childid` of the topic structure is used as device-name.

Reviewer: Does it mean that the `device-name` declared here can be different than the one that is present in the `topic`?

Author: Yes. The device-name is a human readable string. Instead the MQTT topic structure contains the unique child-id.

To speak in C8Y terms:
* the child-id is the external-id, that is used to address the device
* the device name is just a string, e.g. visiable in the UI

See also https://cumulocity.com/guides/reference/smartrest-two/#child-device-creation-101

     Section [thin-egde APIs provided on network](#2-thin-egde-apis-provided-on-network-and-3-associate-consumedprovided-to-child-device-twins) below lists all APIs provided by thin-edge for child-device support.

Example:
```json
{
   "device-name": "My Child-Device",
   "device-type": "type1",
   "configurations": [ // capabilities for feature Configuration Management 
     {
       "type": "foo.conf"
     },
     {
       "type": "bar.conf"
     }
   ]
}
```

Whenever that MQTT message is sent to the thin-edge device:
  1) thin-edge assures the child-device twin exists, and creates it if it does not exist.
  2) each software component that provides an included capability (e.g. the C8Y Configuration Plugin for `configurations`) takes care to define all necessary capabilities to the coresponding cloud's child-device twin (i.E. _supported operations_, and list of provided types as e.g. _configuration types_).

> As result of the child-device creation thin-edge publishs an empty to MQTT topic `tedge/meta/success/<childid>` on successs, or a message to `tedge/meta/failed/<childid>`. The message on `tedge/meta/failed/<childid>` can optional contain a fail reaon as plain text.

Reviewer: ```suggestion
As a result of the child-device creation, thin-edge publishes an empty to MQTT topic `tedge/meta/success/<childid>` on success, or a message to `tedge/meta/failed/<childid>`. The message on `tedge/meta/failed/<childid>` can optionally contain a fail reason as plain text.
```

Author: Meanwhile wording was improved, see https://github.com/thin-edge/thin-edge.io/pull/1236#discussion_r946383052

> As result of the child-device creation thin-edge publishs an empty to MQTT topic `tedge/meta/success/<childid>` on successs, or a message to `tedge/meta/failed/<childid>`. The message on `tedge/meta/failed/<childid>` can optional contain a fail reaon as plain text.

Reviewer 2: Need to document how we handle these failure cases:

1. If thin-edge is not configured to connect to Cumulocity when this message is received, will we do the provisioning after the connect happens?
2. When the device is connected, how do we handle a temporary network blip between the device and Cumulocity 
when this message is received?

Author: Removed from that PR, but to be considered in PR "requirements and use-cases" (#1195).
See more details in comment https://github.com/thin-edge/thin-edge.io/pull/1236#discussion_r955257236

### (2) thin-egde APIs provided on network<br/> and (3) associate consumed/provided to child-device twins

Each thin-edge API that supports child-devices covers both aspects. As of now, the following APIs have child-device support:
   * Measurements, see [Sending a measurements to a child-devices](../tutorials/send-thin-edge-data.md#sending-measurements-to-child-devices)
   * Events, see [Sending an event to a child-device](../tutorials/send-events.md#sending-an-event-for-a-childexternal-device-to-the-cloud)
   * Configuration Management, see [Configuration files for child devices](../references/c8y-configuration-management.md#configuration-files-for-child-devices)

> 

Reviewer: Also, you can link the Alarms doc as well.

Author: added.

- **external-device**:
  some kind of physical device that is connected to the _thin-edge device_ (e.g. a PLC, a fieldbus device, a sensor or any other kind of device)
- **logical child-device**:
  a process running on the _thin-edge device_ that makes use of thin-edge's child-device API to be represented as individual device in the cloud
- **child-device**:
  a _logical child-device_ or an _external-device_ that makes use of thin-edge's child-device API to be represented as individual device in the cloud
- **child-device twin**:
  the entity in the cloud, that represents the child-device; in case of C8Y that is a managed object
- **child-id**:
  a unique string referencing the _external-device_ or _logical child-device_ to the corresponding _child-device twin_
- **child-device agent**:
  that piece of software that makes the glue between the _external-device_ and thin-edge APIs; could be located on the _external-device_ or running as process on the _thin-edge device_

## Principles

Child-device support basically focuses on three aspects:
   1) provision the child-device, i.e. making thin-edge and the cloud aware of the child-device (external or logical) 

>    2) thin-egde APIs provided on network, to enable external-devices APIs to access them

Author: See https://github.com/thin-edge/thin-edge.io/pull/1236#discussion_r956977802

>    3) associate consumed/provided data of a child-device (external or logical) to it's child-device twin in the cloud

Author: What I was trying to say is:
* to (1) 

  the child-device must have a chance to say to thin-edge "hello, I am there", and "I am capable for config man, and logging and softward man"
  
* for (2)

  the child-device must be able to access thin-edge/plugins functionality (e.g. cfg man, logging, software man) via network APIs, instead of e.g. thin-edge local filesystem or  local CLI
  
* for (3)

  the child-device must be able tell thin-edge/plugins which child-device twin in the cloud shall be associated with it's functionality/data (e.g. cfg-man, logging, software man, ... measurements, ...) and thin-edge/plugins have to tell that to the cloud.

>    3) associate consumed/provided data of a child-device (external or logical) to it's child-device twin in the cloud

Author: @didier-wenzek ,  whats about that:

Child-device support basically focuses on three aspects:
   1) the child-device can be provisioned to thin-edge, i.e. making thin-edge and the cloud aware of the child-device (external or logical)
   2) the child-device can access all relevant thin-egde APIs on local network (e.g. for telemetry data, config management, software management, ...)
   3) the child-device can specify for each API call a child-device twin in the cloud, to which all data/operations of the call will be associated

### (1) Child-Device Provisioning

A newly attached child-device (external or logical) must be once provisioned. The provisioning phase of a child-device includes:
  1) creating cloud's child-device twin
  2) declaring all supported capabilities of the child-device to its child-device twin
     (i.E. declaring all _supported operations_ and all _types_ per supported operation, as _config types_, _log types_, ...)
     (i.e. declaring all operations supported by the device like configuration management, software management etc and further metadata per supported operation, like _config types_, _log types_, _software list_ etc)

There are two option to create the cloud's child-device twin and declaring all supported capabilities. 

> On the one hand a cloud's _child-device twin_ can be created upfront (e.g. with some customer-specific cloud-site backend), and also all capabilities can be declared that way to the new twin. Then the child-device on the device site (external or logical) relies on that existing twin.

Author: Removed from that PR, but to be considered in PR "requirements and use-cases" (#1195).
See more details in comment https://github.com/thin-edge/thin-edge.io/pull/1236#discussion_r955257236

     Section [thin-egde APIs provided on network](#2-thin-egde-apis-provided-on-network-and-3-associate-consumedprovided-to-child-device-twins) below lists all APIs provided by thin-edge for child-device support.

Example:
```json
{
   "device-name": "My Child-Device",
   "device-type": "type1",
   "configurations": [ // capabilities for feature Configuration Management 
     {
       "type": "foo.conf"
     },
     {
       "type": "bar.conf"
     }
   ]
}
```

Whenever that MQTT message is sent to the thin-edge device:
  1) thin-edge assures the child-device twin exists, and creates it if it does not exist.
  2) each software component that provides an included capability (e.g. the C8Y Configuration Plugin for `configurations`) takes care to define all necessary capabilities to the coresponding cloud's child-device twin (i.E. _supported operations_, and list of provided types as e.g. _configuration types_).

> As result of the child-device creation thin-edge publishs an empty to MQTT topic `tedge/meta/success/<childid>` on successs, or a message to `tedge/meta/failed/<childid>`. The message on `tedge/meta/failed/<childid>` can optional contain a fail reaon as plain text.

Author: Removed from that PR, but to be considered in PR "requirements and use-cases" (#1195).
See more details in comment https://github.com/thin-edge/thin-edge.io/pull/1236#discussion_r955257236

   * Configuration Management, see [Configuration files for child-devices](../references/c8y-configuration-management.md#configuration-files-for-child-devices)

## User Journeys

The user journeys below outline the behaviour of external-devices, thin-edge and the cloud for the provisioning phase.

#### Personas used in user journeys
* **device-operator**: the one who maintains/operates the device in the field in the shopfloor
   
#### User Journey 1: Connecting an external-device as new child-device, provisioning managed by device
  - device-operator: connects new external-device to the thin-edge device by cable
  - device-operator: powers up the external-device
    - the external-device announces it-self to thin-edge with it's _child-id_ and capabilities (e.g. provided _configuration types_, _log types_, ...)
      Details, see: [Child-Device provisioning](#1-child-device-provisioning) above
    - thin-edge creates the child-device twin in the cloud, based on the _child-id_ reported by the external-evice

      NOTE: Cloud's operation to create the child-device twin is idempotent, so trying to create a child-device twin again that already exist does not harm.

>     - thin-edge declares all capabilities reported by the external-device to the child-device twin (i.E. all _supported operations_ and corresponding types per operations, as _config-types_, _log-types_, ...)

Author: Removed from that PR, but to be considered in PR "requirements and use-cases" (#1195).
See more details in comment https://github.com/thin-edge/thin-edge.io/pull/1236#discussion_r955272147

A newly attached child-device (external or logical) must be once provisioned. The provisioning phase of a child-device includes:
  1) creating cloud's child-device twin
  2) declaring all supported capabilities of the child-device to its child-device twin
     (i.e. declaring all _supported operations_ and all _types_ per supported operation, as _config types_, _log types_, ...)
     (i.e. declaring all operations supported by the device like configuration management, software management etc and further metadata per supported operation, like _config types_, _log types_, _software list_ etc)

There are two options to create the cloud's child-device twin and declaring all supported capabilities:

1. The cloud's _child-device twin_ is created upfront (e.g. with some customer-specific cloud-site backend), and also all capabilities are declared that way to the new twin. Then the child-device on the device site (external or logical) relies on that existing twin.

2. The child-device initiates the device twin creation by announcing its identity and capabilities to thin-edge and then thin-edge gets that child device twin created in the cloud using a cloud mapper.

> 
> ##### Announcing child-device capabilities by the child-device it-self
> 
> To announce it's capabilities the child-device (external or logical) sends an MQTT message to `tedge/meta/<childid>`. The MQTT message is as below:
> 
> ```json
> {
>    "device-name": "<optional name of that child-device>",
>    "device-type": "<optional type of that child-device>",
>    "<capability 1>": <capability specific JSON object>,
>    "<capability 2>": <capability specific JSON object>,
>    // [...]
> }
> ```
> 
> Thereby the fields are as below:
>    * `device-name` is an optional human readble device-name, visible in the cloud. If the field is not contained in the JSON message the `childid` of the topic structure is used as device-name.
>    * `device-type` is an optional device-type string assigned to the cloud's child-device twin. If the field is not contained in the JSON message the value `thin-edge.io-child` is used as device-type string.
>    * each field `<capability i>` represents a capability (e.g. `configurations` for _Configuration management_). The format of each `capability specific JSON object` is specific to the coresponding capability. For details see the documentation for coresponding capabilitie's API (e.g. [Configuration files for child-devices](../references/c8y-configuration-management.md#configuration-files-for-child-devices)). 
>      Section [thin-egde APIs provided on network](#2-thin-egde-apis-provided-on-network-and-3-associate-consumedprovided-to-child-device-twins) below lists all APIs provided by thin-edge for child-device support.
> 
> Example:
> ```json
> {
>    "device-name": "My Child-Device",
>    "device-type": "type1",
>    "configurations": [ // capabilities for feature Configuration Management 
>      {
>        "type": "foo.conf"
>      },
>      {
>        "type": "bar.conf"
>      }
>    ]
> }
> ```
> 
> Whenever that MQTT message is sent to the thin-edge device:
>   1) thin-edge assures the child-device twin exists, and creates it if it does not exist.
>   2) each software component that provides an included capability (e.g. the C8Y Configuration Plugin for `configurations`) takes care to define all necessary capabilities to the coresponding cloud's child-device twin (i.E. _supported operations_, and list of provided types as e.g. _configuration types_).
> 
> In response to the child-device provisioning request, thin-edge publishes an empty resposne to MQTT topic `tedge/meta/success/<child-id>` on success, or a failure message to `tedge/meta/failed/<child-id>` in case of a failure. The failure message on `tedge/meta/failed/<childid>` can be optional.
> 

Author: Too much technical detail about MQTT topics and payload, without any investigation or concept about (child-) device data-model and inventory data-model. In addition that PR is the wrong scope for details about provisioning.

Content should be removed here. Nevertheless it should be considered where focusing on provisioning.

```suggestion
```

### (2) thin-edge APIs provided over the network<br/> and (3) associate consumed/provided data to child-device twins

Each thin-edge API that supports child-devices covers both aspects. As of now, the following APIs have child-device support:
   * Measurements, see [Sending a measurements to a child-devices](../tutorials/send-thin-edge-data.md#sending-measurements-to-child-devices)
   * Events, see [Sending an event to a child-device](../tutorials/send-events.md#sending-an-event-for-a-childexternal-device-to-the-cloud)
   * Alarms, see [Raising an alarm to a child-device](../tutorials/raise-alarm.md#raising-alarms-from-child-devices)
   * Configuration Management, see [Configuration files for child-devices](../references/c8y-configuration-management.md#configuration-files-for-child-devices)

> ## User Journeys
> 
> The user journeys below outline the behaviour of external-devices, thin-edge and the cloud during the provisioning phase.
> 
> #### Personas used in user journeys
> * **device-operator**: the one who maintains/operates the device in the field in the shopfloor
>    
> #### User Journey 1: Connecting an external-device as new child-device, provisioning managed by device
>   - device-operator: connects new external-device to the thin-edge device by cable
>   - device-operator: powers up the external-device
>     - the external-device announces it-self to thin-edge with it's _child-id_ and capabilities (e.g. provided _configuration types_, _log types_, ...)
>       Details, see: [Child-Device provisioning](#1-child-device-provisioning) above
>     - thin-edge creates the child-device twin in the cloud, based on the _child-id_ reported by the external-evice
> 
>       NOTE: Cloud's operation to create the child-device twin is idempotent, so trying to create a child-device twin again that already exist does not harm.
>     - thin-edge declares all capabilities reported by the external-device to the child-device twin (i.E. all _supported operations_ and corresponding types per operations, as _config-types_, _log-types_, ...)
>   - from that point the new external-device is ready in operation
> 
> #### User Journey 2: Connecting an external-device as new child-device, provisioning managed by a cloud-site backend
>   - device-operator: creates the child-device twin in the cloud using some customer-specific backend; as _child-id_ the device-operator uses some UID that is known by the external-device (e.g. the external-device's serial number or network address)
>   - device-operator: declares all capabilities to the child-device twin (i.E. all _supported operations_ and corresponding types per operations, as _config-types_, _log-types_, ...)
>   - device-operator: connects new external-device to the thin-edge device by cable
>   - device-operator: powers up the external-device
>     - the external-device announces it-self to thin-edge with it's _child-id_
>     - thin-edge tries to created the child-device twin in the cloud, even though that twin was already created upfront
> 
>       NOTE: Cloud's operation to create the child-device twin is idempotent, so trying to create a child-device twin that already exist does not harm.
> 
>     - thin-edge retrieves all declared capabilities from the child-device twin (e.g. provided _configuration types_, _log types_, ...) and send as response to the external-device
> 
>   - from that point the new external-device is ready in operation
> 

Author: User journeys should be removed from that PR, but applied to PR "requirements and use-cases" (https://github.com/thin-edge/thin-edge.io/pull/1195).

Do not forget to consider user journeys pending comments:
  * https://github.com/thin-edge/thin-edge.io/pull/1236#discussion_r944156805
  * https://github.com/thin-edge/thin-edge.io/pull/1236#discussion_r947026933
  * https://github.com/thin-edge/thin-edge.io/pull/1236#discussion_r944161015
  * https://github.com/thin-edge/thin-edge.io/pull/1236#discussion_r944182229
  * https://github.com/thin-edge/thin-edge.io/pull/1236#discussion_r947029394
  * https://github.com/thin-edge/thin-edge.io/pull/1236#discussion_r944198937
  * https://github.com/thin-edge/thin-edge.io/pull/1236#discussion_r947033546
  
```suggestion

```

# Child-Device Support

Child-Device support allows to connect devices (e.g. sensors, actors, PLCs, any other kind of device) to the thin-edge device, to represent those devices in the cloud. Furthermore processes running on the thin-edge device can make use of child-device support to appear as a kind of a logical device in the cloud.

## Terminology
- **thin-edge device**:
  the physical device thin-edge is running on
- **external-device**:
  some kind of physical device that is connected to the _thin-edge device_ (e.g. a PLC, a fieldbus device, a sensor or any other kind of device)
- **child-device twin**:
  the entity in the cloud, that represents the child-device; in case of C8Y that is a managed object
- **child-id**:
  a string uniquely referencing the _external-device_ and the corresponding _child-device twin_
- **child-device agent**:
  that piece of software that makes the glue between the _external-device_ and thin-edge APIs; could be located on the _external-device_ or running as process on the _thin-edge device_

> 

Author: So far for the scope of that PR (Child-Device Support for C8Y Config Plugin) there is no need to introduce "external child device agent" and "local child device agent". However for the overall scope it might make sense. So I will consider that comment again in PR "requirements and use-cases" (#1195).

## Principles

Child-device support basically focuses on two aspects:
   1) enable external-devices to access thin-edge APIs via network

>    2) equip thin-edge APIs to associate consumed/provided data with a child-device twin in the cloud

Author: That part was changed meanwhile.

See also https://github.com/thin-edge/thin-edge.io/pull/1236/files#r957001530

> Each thin-edge API that supports child-devices applies to both aspects. As of now APIs as below have child-device support:

Author: I merged that suggestion manually in my meanwhile locally changes version.

   * Measurements, see [Sending a measurements to a child-devices](../tutorials/send-thin-edge-data.md#sending-measurements-to-child-devices)
   * Events, see [Sending an event to a child-device](../tutorials/send-events.md#sending-an-event-for-a-childexternal-device-to-the-cloud)
   * Configuration Management, see [Configuration files for child devices](../references/c8y-configuration-management.md#configuration-files-for-child-devices)

## Child-Device Provisioning

The provisioning phase of a child-device includes:
  1) creating the child-device twin in the cloud

>   2) declaring all supported capabilities on the child-device twin in the cloud

Author: I merged that suggestion manually in my meanwhile locally changes version.

>      (i.E. declaring all _supported operations_ and all _types_ per supported operation, as _config types_, _log types_, ...) 

Author: I merged that suggestion manually in my meanwhile locally changes version.

The user journeys below outline the behaviour of external-devices, thin-edge and the cloud for the provisioning phase.

#### Personas used in user journeys

> * **device-operator**: the one who maintains/operates the device in the field in the shopfloor

Author: Removed from that PR, but to be considered in PR "requirements and use-cases" (#1195).
See more details in comment https://github.com/thin-edge/thin-edge.io/pull/1236#discussion_r955272147

> * **device-operator**: the one who maintains/operates the device in the field in the shopfloor

Author: Removed from that PR, but to be considered in PR "requirements and use-cases" (#1195).
See more details in comment https://github.com/thin-edge/thin-edge.io/pull/1236#discussion_r955272147

   
#### User Journey 1: Connecting an external-device as new child-device, provisioning managed by device
  - device-operator: connects new external-device to the thin-edge device by cable
  - device-operator: powers up the external-device

>     - the external-device announces it-self to thin-edge with it's _child-id_ and capabilities (e.g. provided _configuration types_, _log types_, ...)

Author: Removed from that PR, but to be considered in PR "requirements and use-cases" (#1195).
See more details in comment https://github.com/thin-edge/thin-edge.io/pull/1236#discussion_r955272147

      Details, see: [_configuration types_](https://github.com/thin-edge/thin-edge.io/blob/cdab5683de9f9e0f34fa42a094ac399f6dbdc1e3/docs/src/references/c8y-configuration-management.md#managing-supported-configuration-list-of-child-devices)
    - thin-edge creates the child-device twin in the cloud, based on the _child-id_ reported by the external-evice

>       NOTE: Cloud's operation to create the child-device twin is idempotent, so trying to create a child-device twin again that already exist does not harm.

Reviewer 2: ```suggestion
      Assumption: Cloud's operation to create the child-device twin is idempotent, so trying to create a child-device twin again that already exist does not harm.
```
As that's something that thin-edge expects from the cloud rather than something we can state to be true for all clouds (even though it's true for Cumulocity).

Author: Removed from that PR, but to be considered in PR "requirements and use-cases" (#1195).
See more details in comment https://github.com/thin-edge/thin-edge.io/pull/1236#discussion_r955272147

    - thin-edge declares all capabilities reported by the external-device to the child-device twin (i.E. all _supported operations_ and corresponding types per operations, as _config-types_, _log-types_, ...)
  - from that point the new external-device is ready in operation

#### User Journey 2: Connecting an external-device as new child-device, provisioning managed by a cloud-site backend
  - device-operator: creates the child-device twin in the cloud using some customer-specific backend; as _child-id_ the device-operator uses some UID that is known by the external-device (e.g. the external device's serial number or network address)
  - device-operator: connects new external-device to the thin-edge device by cable
  - device-operator: powers up the external-device

>     - the external-device announces it-self to thin-edge with it's _child-id_ and capabilities (e.g. provided _configuration types_, _log types_, ...)

Reviewer 2: Why are we expecting the device to send the capabilities as per this flow? The device twin and its capabilities are provisioned directly on the cloud, right?

Author: Removed from that PR, but to be considered in PR "requirements and use-cases" (#1195).
See more details in comment https://github.com/thin-edge/thin-edge.io/pull/1236#discussion_r955272147

>     - thin-edge tries to created the child-device twin in the cloud, even though that twin was already created upfront.

Reviewer 2: This device creation attempt should be replaced with a device existence check here, right? If the device exists already, thin-edge will have to fetch all the device capabilities from the cloud so that it can keep all that information persisted on its file system.

Author: Removed from that PR, but to be considered in PR "requirements and use-cases" (#1195).
See more details in comment https://github.com/thin-edge/thin-edge.io/pull/1236#discussion_r955272147

      NOTE: Cloud's operation to create the child-device twin is idempotent, so trying to create a child-device twin that already exist does not harm.
    - thin-edge declares all capabilities reported by the external-device to the child-device twin (i.E. all _supported operations_ and corresponding types per operations, as _config-types_, _log-types_, ...)
  - from that point the new external-device is ready in operation

> 

Reviewer 2: This flow looks exactly the same as in the previous user journey where the device initiates the provisioning. How can they be the same? When the device twin is provisioned in the cloud first, my assumption was that the device will be syncing its capabilities and other metadata from the cloud. Wasn't that the whole point of this flow? To offload all the device provisioning knowledge from the device to the cloud?

Reviewer 3: This flow is really confusing and lacking too many details to be implemented.

Author: Removed from that PR, but to be considered in PR "requirements and use-cases" (#1195).
See more details in comment https://github.com/thin-edge/thin-edge.io/pull/1236#discussion_r955272147

## Principles

Child-device support basically focuses on three aspects:
   1) provision the child-device, i.e. making thin-edge and the cloud aware of the child-device (external or logical) 

>    2) thin-egde APIs provided on network, to enable external APIs to access them

Reviewer 2: ```suggestion
   2) thin-edge APIs exposed over the network, to enable external child devices to access them
```

Author: Please see https://github.com/thin-edge/thin-edge.io/pull/1236#discussion_r957001530

   3) associate consumed/provided data of a child-device (external or logical) to it's child-device twin in the cloud

### (1) Child-Device Provisioning

A newly attached child-device (external or logical) must be once provisioned. The provisioning phase of a child-device includes:
  1) creating cloud's child-device twin
  2) declaring all supported capabilities of the child-device to its child device twin

>      (i.E. declaring all _supported operations_ and all _types_ per supported operation, as _config types_, _log types_, ...)

Reviewer 2: This line is redundant with the next line. I'd just keep the next one.

Author: Removed from that PR, but to be considered in PR "requirements and use-cases" (#1195).
See more details in comment https://github.com/thin-edge/thin-edge.io/pull/1236#discussion_r955257236

     (i.e. declaring all operations supported by the device like configuration management, software management etc and further metadata per supported operation, like _config types_, _log types_, _software list_ etc)

There are two option to create the cloud's child-device twin and declaring all supported capabilities. 

On the one hand a cloud's _child-device twin_ can be created upfront (e.g. with some customer-specific cloud-site backend), and also all capabilities can be declared that way to the new twin. Then the child-device on the device site (external or logical) relies on that existing twin.

> On the other hand a child-device (external or logical) can announce it's capabilities to thin-edge to provoke thin-edge creating the cloud's child-device twin and declaring all capabilities on that child-device twin. 

Reviewer 2: ```suggestion
2. The child-device initiates the device twin creation by announcing its identity and capabilities to thin-edge and then thin-edge gets that child device twin created in the cloud using a cloud mapper.
```

Author: Suggested wording applied.

     Section [thin-egde APIs provided on network](#2-thin-egde-apis-provided-on-network-and-3-associate-consumedprovided-to-child-device-twins) below lists all APIs provided by thin-edge for child-device support.

Example:
```json
{
   "device-name": "My Child-Device",
   "device-type": "type1",
   "configurations": [ // capabilities for feature Configuration Management 
     {
       "type": "foo.conf"
     },
     {
       "type": "bar.conf"
     }
   ]
}
```

Whenever that MQTT message is sent to the thin-edge device:
  1) thin-edge assures the child-device twin exists, and creates it if it does not exist.
  2) each software component that provides an included capability (e.g. the C8Y Configuration Plugin for `configurations`) takes care to define all necessary capabilities to the coresponding cloud's child-device twin (i.E. _supported operations_, and list of provided types as e.g. _configuration types_).

> As result of the child-device creation thin-edge publishs an empty to MQTT topic `tedge/meta/success/<childid>` on successs, or a message to `tedge/meta/failed/<childid>`. The message on `tedge/meta/failed/<childid>` can optional contain a fail reaon as plain text.

Reviewer 2: ```suggestion
In response to the child-device provisioning request, thin-edge publishes an empty resposne to MQTT topic `tedge/meta/success/<child-id>` on success, or a failure message to `tedge/meta/failed/<child-id>` in case of a failure. The failure message on `tedge/meta/failed/<childid>` can be optional.
```

Author: Suggested applied.

> ### (2) thin-egde APIs provided on network<br/> and (3) associate consumed/provided to child-device twins

Reviewer 2: ```suggestion
### (2) thin-edge APIs provided over the network<br/> and (3) associate consumed/provided data to child-device twins
```
